Sources/Basic/OutputByteStream.swift

```diff
@@ -55,27 +60,86 @@ public class OutputByteStream: TextOutputStream {
         return buffer.count
     }
 
-    /// The contents of the output stream.
-    ///
-    /// This method implicitly flushes the stream.
-    public final var bytes: ByteString {
-        flush()
-        return ByteString(self.buffer)
+    /// Currently available buffer size.
+    private var availableBufferSize: Int {
+        return buffer.capacity - buffer.count
     }
-    
+
+     /// Clears the buffer maintaining current capacity.
+    private func clearBuffer() {
+        buffer.removeAll(keepingCapacity: true)
+    }
+
     // MARK: Data Output API
 
     public final func flush() {
+        writeImpl(buffer)
+        clearBuffer()
+        flushImpl()
+    }
+
+    func flushImpl() {
         // Do nothing.
     }
 
+    func writeImpl<C: Collection where C.Iterator.Element == UInt8>(_ bytes: C) {
```

Comment:
Oh, I missed this.

I don't think we want `writeImpl` to be generic, that isn't likely to be efficient (it would require compiler to specialize across a class hierarchy, I don't know if it can do that (cc @atrick) . I think we want it to take an `UnsafeBufferPointer` under the expectation that we are usually sending our own buffer down to it, and will be able to get one cheaply.

I'm not sure how this impacts `write` for collections...